Rienafairefr's templating branch updated to current master

jimschubert opened the pull request:
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`~~, `3.4.x`, `4.0.x`~~. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

master merge and Locale.ROOT fix to get #690 to cleanly merge into master.

For any discussions, please continue the thread at #690.

jimschubert commented:
@OpenAPITools/openapi-generator-core-team Anyone able to review for potential breaking changes concerns? I'd like to get the functionality into 4.0.0 major release.

wing328 commented on modules/openapi-generator-cli/src/main/java/org/openapitools/codegen/cmd/Generate.java:
```diff
@@ -70,6 +70,10 @@
     private String templateDir;
 
+    @Option(name = {"-e", "--engine"}, title = "templating engine",
+        description = "templating engine, for now \"mustache\" and \"handlebars\" are supported")
```
Minor suggestion: make it clear that mustache is the default and handlebar is beta only (not officially support)

wing328 commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenConstants.java:
```diff
@@ -184,6 +184,9 @@
     public static final String DOTNET_FRAMEWORK_DESC = "The target .NET framework version.";
 
+    public static final String TEMPLATING_ENGINE = "templatingEngine";
+    public static final String TEMPLATING_ENGINE_DESC = "The templating engine plugin to use";
```
Minor suggestion similar to https://github.com/OpenAPITools/openapi-generator/pull/2657/files#r275162756

jimschubert commented:
Breaking change with b5f2c2a, avoiding confusing package location.

jimschubert commented:
@wing328 thanks for the review. I added some clarity around those descriptions. I've also relocated package `org.openapitools.codegen.mustache` to `org.openapitools.codegen.templating.mustache`, and implemented a `startsWith` template helper as requested in #690.

While writing the unit test for this handlebars helper, I noticed the tests I had previously written for the mustache helpers no longer exist. Do you recall why they were removed, and if not… any objection if I add those tests back in? I looked in history and it seems they may have been removed mistakenly during a package rename.

wing328 commented:
> While writing the unit test for this handlebars helper, I noticed the tests I had previously written for the mustache helpers no longer exist. Do you recall why they were removed, and if not… any objection if I add those tests back in? I looked in history and it seems they may have been removed mistakenly during a package rename.

No objection at all and I don't recall removing those tests.

We definitely welcome more tests for this project.

jimschubert commented:
Cool. I'll create an issue and sign it to myself as a reminder.

jimschubert commented:
Task for tracking adding back the tests is #2682
